src/context/CharacterContext.js

```diff
@@ -0,0 +1,48 @@
+import { createContext, useContext, useEffect, useState } from 'react';
+import { fetchCharacterById, fetchCharacters } from '../services/fetchRandM';
+
+const CharacterContext = createContext();
+
+const CharacterProvider = ({ children }) => {
+  const [characters, setCharacters] = useState([]);
+  const [character, setCharacter] = useState({});
+  const [loading, setLoading] = useState(true);
+
+  const getCharacterById = async (id) => {
+    const thisCharacter = characters.find((val) => val.id === Number(id));
+    if (!thisCharacter) {
+      setLoading(true);
+      const fetchData = async () => {
+        const data = await fetchCharacterById(id);
+        setCharacter(data);
+      };
+      fetchData();
+    } else {
+      setCharacter(thisCharacter);
+    }
+    setLoading(false);
+  };
```

Comment:
We could also place this into a useEffect in context with the id and loading as dependencies.